### PR TITLE
TEC-3476: Add option to send "PUT" calls to the EA server endpoints

### PR DESCRIPTION
**Pull request by mitogh**
- Allow making PUT requests to the EA server
- Allow updating a registered import record from EA back to the EA server.

Ticket: [TEC-3476]

[TEC-3476]: https://moderntribe.atlassian.net/browse/TEC-3476

### CODE REVIEW

File: src/Tribe/Aggregator/API/Import.php [hunk 1]
```diff
@@ -107,9 +107,19 @@ public function get( $import_id, $data = array() ) {
 	 * @return stdClass|WP_Error
 	 */
 	public function create( $args ) {
-		$response = $this->service->post_import( $args );
+		return $this->service->post_import( $args );
+	}
 
-		return $response;
+	/**
```
Comment (tr1b0t) [inline src/Tribe/Aggregator/API/Import.php:113]: * :no_entry_sign: Doc comment for parameter "$import_id" missing
* :no_entry_sign: Doc comment for parameter "$args" missing

<sub>hash: 7bdf5347d10666c43136740c0818a700</sub>

File: src/Tribe/Aggregator/API/Import.php [hunk 2]
```diff
@@ -107,9 +107,19 @@ public function get( $import_id, $data = array() ) {
...
+	 * Update the details of an existing import
+	 *
+	 * @param $import_id
```
Comment (tr1b0t) [inline src/Tribe/Aggregator/API/Import.php:116]: :no_entry_sign: Missing parameter name

<sub>hash: 2df0e292677ccc70a658f7c6d5d3c9d8</sub>

File: src/Tribe/Aggregator/API/Import.php [hunk 3]
```diff
@@ -107,9 +107,19 @@ public function get( $import_id, $data = array() ) {
...
+	 * @param $args
```
Comment (tr1b0t) [inline src/Tribe/Aggregator/API/Import.php:117]: :no_entry_sign: Missing parameter name

<sub>hash: bb1b38abd678686cb4fa868769a70d97</sub>

File: src/Tribe/Aggregator/Service.php [hunk 4]
```diff
@@ -526,6 +526,36 @@ public function post_import( $args ) {
 		return $this->post( 'import', $args );
 	}
 
+	public function update_import( $import_id, $args ) {
```
Comment (tr1b0t) [inline src/Tribe/Aggregator/Service.php:529]: :no_entry_sign: Missing doc comment for function update_import()

<sub>hash: acab1b7104194a15e5e4d15768cbb5d9</sub>

File: src/Tribe/Aggregator/Service.php [hunk 5]
```diff
@@ -526,6 +526,36 @@ public function post_import( $args ) {
...
+		$api = $this->api();
+
+		// if the user doesn't have a license key, don't bother hitting the service
```
Comment (tr1b0t) [inline src/Tribe/Aggregator/Service.php:532]: :no_entry_sign: Inline comments must end in full-stops, exclamation marks, or question marks

<sub>hash: ac4c751de21ae3443f71dfeac2d850c1</sub>

File: src/Tribe/Aggregator/Service.php [hunk 6]
```diff
@@ -526,6 +526,36 @@ public function post_import( $args ) {
...
+		 *
+		 * @since TBD
+		 *
+		 * @param  bool|string $pue_key PUE key
+		 * @param  array       $args    Arguments to queue the import
+		 * @param  self        $record  Which record we are dealing with
+		 */
+		$licenses = apply_filters( 'tribe_aggregator_service_put_pue_licenses', [], $args, $this );
+
+		// If we have a key we add that to the Arguments
```
Comment (tr1b0t) [inline src/Tribe/Aggregator/Service.php:548]: :no_entry_sign: Inline comments must end in full-stops, exclamation marks, or question marks

<sub>hash: 9e961ad679d2852132624af1b72dd34b</sub>

File: src/Tribe/Aggregator/Service.php [hunk 7]
```diff
@@ -526,6 +526,36 @@ public function post_import( $args ) {
...
+		if ( ! empty( $licenses ) ) {
+			$args['licenses'] = $licenses;
+		}
+
+		return $this->post( "import/{$import_id}", [
```
Comment (tr1b0t) [inline src/Tribe/Aggregator/Service.php:553]: * :no_entry_sign: Opening parenthesis of a multi-line function call must be the last content on the line
* :no_entry_sign: Only one argument is allowed per line in a multi-line function call

<sub>hash: c9aa13737ef44f105f6e276f104de57b</sub>

File: src/Tribe/Aggregator/Service.php [hunk 8]
```diff
@@ -526,6 +526,36 @@ public function post_import( $args ) {
...
+			'body' => $args,
```
Comment (tr1b0t) [inline src/Tribe/Aggregator/Service.php:554]: :warning: Array double arrow not aligned correctly; expected 3 space(s) between "'body'" and double arrow, but found 1.

<sub>hash: 7aeb0133b5a00cfa05683d602d7bbe6d</sub>

File: src/Tribe/Aggregator/Service.php [hunk 9]
```diff
@@ -526,6 +526,36 @@ public function post_import( $args ) {
...
+			'method' => 'PUT',
+		] );
```
Comment (tr1b0t) [inline src/Tribe/Aggregator/Service.php:556]: :no_entry_sign: Closing parenthesis of a multi-line function call must be on a line by itself

<sub>hash: 953e80ad5b4de9193534b20d16aeb5e0</sub>

File: readme.txt [hunk 10]
```diff
@@ -1,8 +1,8 @@
 === The Events Calendar ===
```
Comment (lucatume) [inline readme.txt:1]: The changes to the readme are not in scope with the issue linked in the description.
Pull them out of this PR and PR them separately, specifying the issue they are related to.
Or update the description to link to the issue the changes are for.

File: src/Tribe/Aggregator/API/Import.php [hunk 11]
```diff
@@ -107,9 +107,19 @@ public function get( $import_id, $data = array() ) {
...
+	 *
+	 * @return object|stdClass|string|WP_Error
```
Comment (lucatume) [inline src/Tribe/Aggregator/API/Import.php:119]: Add doc entry.

File: src/Tribe/Aggregator/API/Import.php (hunk 3 above)
Comment (lucatume) [inline src/Tribe/Aggregator/API/Import.php:117]: Add doc entry.

Comment (lucatume) [reply src/Tribe/Aggregator/API/Import.php:117]: Add doc entry.

File: src/Tribe/Aggregator/API/Import.php (hunk 2 above)
Comment (lucatume) [reply src/Tribe/Aggregator/API/Import.php:116]: Add doc entry.

File: src/Tribe/Aggregator/Service.php [hunk 12]
```diff
@@ -526,6 +526,36 @@ public function post_import( $args ) {
...
+	public function update_import( $import_id, $args ) {
+		$api = $this->api();
+
+		// if the user doesn't have a license key, don't bother hitting the service
+		if ( is_wp_error( $api ) ) {
+			return $api;
+		}
+
+		/**
+		 * Allows filtering to add a PUE key to be passed to the EA service
```
Comment (lucatume) [inline src/Tribe/Aggregator/Service.php:538]: There is something wonky in the wording of this. And the full-stop at the end is missing.

File: src/Tribe/Aggregator/Service.php [hunk 13]
```diff
@@ -526,6 +526,36 @@ public function post_import( $args ) {
...
+		 *
+		 * @since TBD
+		 *
+		 * @param  bool|string $pue_key PUE key
```
Comment (lucatume) [inline src/Tribe/Aggregator/Service.php:542]: For this and below: full-stop at the end of the comment.

File: readme.txt (hunk 10 above)
Comment (bordoni) [reply readme.txt:1]: I think this was pulled from Master. Made by Z.

**Comment by lucatume** [comment]
I've re-started the tests as Travis CI seemed stuck.

**Comment by mitogh** [comment]
Oh is actually still there, we are reusing the `post` method to make a PUT call.

https://github.com/moderntribe/the-events-calendar/blob/d25ed5fe3d87f63366748c13aec7e4a132bb4431/src/Tribe/Aggregator/Service.php#L539-L571

**Comment by lucatume** [comment]
> Oh is actually still there, we are reusing the `post` method to make a PUT call.

You're one sly fox.

**Pull request merged**